File: components/volumeRender/textures/typescript/colormapTextures.ts
/**
 * yuki-koyama/tinycolormap
 * @link https://github.com/yuki-koyama/tinycolormap
 *
 * tinycolormap.hpp
 * @link https://github.com/yuki-koyama/tinycolormap/blob/d75a66dd4ba8469eadadb953a843aa2cc4dc357c/include/tinycolormap.hpp#L199
 */

import * as THREE from "three";

import {
    GetCividisColor,
    GetCubehelixColor,
    GetGithubColor,
    GetGrayColor,
    GetHeatColor,
    GetHotColor,
    GetInfernoColor,
    GetJetColor,
    GetMagmaColor,
    GetParulaColor,
    GetPlasmaColor,
    GetTurboColor,
    GetViridisColor,
} from "./colormap";

function GetColor(x: number, type: string): THREE.Color {
    let color: THREE.Color;

    switch (type) {
        case "parula":
            color = GetParulaColor(x);
            break;
        case "heat":
            color = GetHeatColor(x);
            break;
        case "jet":
            color = GetJetColor(x);
            break;
        case "turbo":
            color = GetTurboColor(x);
            break;
        case "hot":
            color = GetHotColor(x);
            break;
        case "gray":
            color = GetGrayColor(x);
            break;
        case "magma":
            color = GetMagmaColor(x);
            break;
        case "inferno":
            color = GetInfernoColor(x);
            break;
        case "plasma":
            color = GetPlasmaColor(x);
            break;
        case "viridis":
            color = GetViridisColor(x);
            break;
        case "cividis":
            color = GetCividisColor(x);
            break;
        case "github":
            color = GetGithubColor(x);
            break;
        case "cubehelix":
            color = GetCubehelixColor(x);
            break;
        default:
            color = new THREE.Color(0xff55dd);
            break;
    }

    return color;
}

/**
 *
 * @param width
 * @param height
 * @param type parula, heat, jet, turbo, hot, gray, magma, inferno, plasma, viridis, cividis, github, cubehelix
 * @returns
 */
export function GetTexture(width: number, height: number, type: string) {
    var size = width * height;
    var data = new Uint8Array(size * 4);

    for (let i = 0; i < height; i++) {
        for (let j = 0; j < width; j++) {
            var color = GetColor(j / (width - 1), type).clone();
            var stride = (i * width + j) * 4;
            data[stride] = color.r * 255;
            data[stride + 1] = color.g * 255;
            data[stride + 2] = color.b * 255;
            data[stride + 3] = 1;
        }
    }
    const texture = new THREE.DataTexture(
        data,
        width,
        height,
        THREE.RGBAFormat,
        THREE.UnsignedByteType,
    );
    texture.magFilter = THREE.NearestFilter;
    texture.needsUpdate = true;

    return texture;
}

export const cmtextures: {[index: string]: THREE.DataTexture} = {
    parula: GetTexture(256, 1, "parula"),
    heat: GetTexture(256, 1, "heat"),
    jet: GetTexture(256, 1, "jet"),
    turbo: GetTexture(256, 1, "turbo"),
    hot: GetTexture(256, 1, "hot"),
    gray: GetTexture(256, 1, "gray"),
    magma: GetTexture(256, 1, "magma"),
    inferno: GetTexture(256, 1, "inferno"),
    plasma: GetTexture(256, 1, "plasma"),
    viridis: GetTexture(256, 1, "viridis"),
    cividis: GetTexture(256, 1, "cividis"),
    github: GetTexture(256, 1, "github"),
    cubehelix: GetTexture(256, 1, "cubehelix")
}

File: src/volumeRender/textures/python/colormapTextures.py
# coding: utf-8

"""
yuki-koyama/tinycolormap
@link https://github.com/yuki-koyama/tinycolormap

tinycolormap.hpp
@link https://github.com/yuki-koyama/tinycolormap/blob/d75a66dd4ba8469eadadb953a843aa2cc4dc357c/include/tinycolormap.hpp#L199
"""

import numpy as np
import cv2

import colormap

def GetColor(x, type):
    color = np.array([0, 0, 0])

    if (type == "parula"):
        color = colormap.parula.GetParulaColor(x)
    elif (type == "heat"):
        color = colormap.heat.GetHeatColor(x)
    elif (type == "jet"):
        color = colormap.jet.GetJetColor(x)
    elif (type == "turbo"):
        color = colormap.turbo.GetTurboColor(x)
    elif (type == "hot"):
        color = colormap.hot.GetHotColor(x)
    elif (type == "gray"):
        color = colormap.gray.GetGrayColor(x)
    elif (type == "magma"):
        color = colormap.magma.GetMagmaColor(x)
    elif (type == "inferno"):
        color = colormap.inferno.GetInfernoColor(x)
    elif (type == "plasma"):
        color = colormap.plasma.GetPlasmaColor(x)
    elif (type == "viridis"):
        color = colormap.viridis.GetViridisColor(x)
    elif (type == "cividis"):
        color = colormap.cividis.GetCividisColor(x)
    elif (type == "github"):
        color = colormap.githubColor.GetGithubColor(x)
    elif (type == "cubehelix"):
        color = colormap.cubeHelix.GetCubehelixColor(x)
    else:
        color = np.array([0xff, 0x55, 0xdd])
        color = color / 0xff

    return color

"""
@param width
@param height
@param type parula, heat, jet, turbo, hot, gray, magma, inferno, plasma, viridis, cividis, github, cubehelix
@returns
"""
def GetTexture(width, height, type, filename="colormap.png"):
    # data = np.zeros((height, width, 4), np.uint8)
    data = np.zeros((height, width, 3), np.uint8)

    for i in range(0, height):
        for j in range(0, width):
            color = GetColor(j / (width - 1), type)
            color = color * 255
            # color = np.append(color, 1)
            data[i][j] = color

    cv2.imwrite(filename, data)


# TODO: python colormapTextures.py
if __name__ == "__main__":
    width = 256
    height = 1
    colormap_type = ["parula", "heat", "jet", "turbo", "hot", "gray", "magma", "inferno", "plasma", "viridis", "cividis", "github", "cubehelix"]
    # colormap_type = ["default"]
    for c in colormap_type:
        GetTexture(width, height, c, "./../img/cm_" + c + ".png")